hwilib/devices/jade.py

```diff
@@ -0,0 +1,323 @@
+    List,
+    Union
+)
+from ..descriptor import PubkeyProvider
+from ..hwwclient import HardwareWalletClient
+from ..errors import (
+    ActionCanceledError,
+    BadArgumentError,
+    DeviceConnectionError,
+    DeviceFailureError,
+    UnavailableActionError,
+    common_err_msgs,
+    handle_errors
+)
+from ..common import (
+    AddressType,
+    Chain,
+)
+from ..key import (
+    ExtendedKey,
+    parse_path
+)
+from ..psbt import PSBT
+from ..tx import (
+    CTransaction
+)
+from .._script import (
+    is_p2sh,
+    is_p2wpkh,
+    is_p2wsh,
+    is_witness,
+)
+
+import logging
+import os
+
+JADE_VENDOR_ID = 0x10c4
+JADE_DEVICE_ID = 0xea60
+
+py_enumerate = enumerate # To use the enumerate built-in, since the name is overridden below
+
+def jade_exception(f):
+    def func(*args, **kwargs):
+        try:
+            return f(*args, **kwargs)
+        except ValueError as e:
+            raise BadArgumentError(str(e))
+        except JadeError as e:
+            if e.code == -32000:  # CBOR_RPC_USER_CANCELLED
+                raise ActionCanceledError('{} canceled by user'.format(f.__name__))
+            elif e.code == -32602:  # CBOR_RPC_BAD_PARAMETERS
+                raise BadArgumentError(e.message)
+            elif e.code == -32603: # CBOR_RPC_INTERNAL_ERROR
+                raise DeviceFailureError(e.message)
+            elif e.code == -32002:  # CBOR_RPC_HW_LOCKED
+                raise DeviceConnectionError('Device is locked')
+            elif e.code == -32003:  # CBOR_RPC_NETWORK_MISMATCH
+                raise DeviceConnectionError('Network/chain selection error')
+            elif e.code in [-32600, -32601, -32001]:  # CBOR_RPC_INVALID_REQUEST, CBOR_RPC_UNKNOWN_METHOD, CBOR_RPC_PROTOCOL_ERROR
+                raise DeviceConnectionError('Messaging/communiciation error')
+            else:
+                raise e
+    return func
+
+# This class extends the HardwareWalletClient for Blockstream Jade specific things
+class JadeClient(HardwareWalletClient):
+
+    NETWORKS = {Chain.MAIN: 'mainnet',
+                Chain.TEST: 'testnet',
+                Chain.REGTEST: 'regtest'}
+
+    def _network(self):
+        return JadeClient.NETWORKS.get(self.chain, 'mainnet')
+
+    ADDRTYPES = {AddressType.PKH: 'pkh(k)',
+                 AddressType.WPKH: 'wpkh(k)',
+                 AddressType.SH_WPKH: 'sh(wpkh(k))'}
+
+    @staticmethod
+    def _convertAddrType(addrType):
+        return JadeClient.ADDRTYPES[addrType]
+
+    def __init__(self, path: str, password: str = '', expert: bool = False) -> None:
+        super(JadeClient, self).__init__(path, password, expert)
+        self.jade = JadeAPI.create_serial(path)
+        self.jade.connect()
+
+        # Push some host entropy into jade
+        self.jade.add_entropy(os.urandom(32))
+
+        # Do the PIN thing if required
+        # NOTE: uses standard 'requests' networking to connect to blind pinserver
+        while not self.jade.auth_user(self._network()):
```

Comment:
Is this interactive (i.e. user enters something on the host computer)? If so, is this required?

Reply:
The user enters the pin on the hww.  The host then mediates between the hww and the remote blind pin server to unlock the keys on the hw.  
It is necessary in a 'real-world' situation to decrypt the users wallet/keys.

With a debug build of the fw [ie in a CI test] we can skip this step and poke a mnemonic or a seed directly into the hw.

Reply:
To clarify further, once the hww is unlocked (whether via pinserver or via 'poke in seed') 'auth_user' is a near no-op which returns immediately without user action / network activity.

So when I am manually testing, the first time I 'enumerate' I may have to unlock the Jade - but for the rest of my testing this call is a no-op (until/unless Jade is turned off).
Usually I manually inject a known seed into jade before I start testing, so a) I don't get interrupted with having to unlock it mid-test, and b) I get known/deterministic addresses/signatures etc for testing.

If/when we do get a sim working, that's what we'd aim to do.

Reply:
What is the "blind pin server"?

Reply:
Ok, well in a real-world situation the keys on a Jade hww cannot be decrypted without this interaction with a remote server.
( It can work offline once 'unlocked' - but the initial step to unlock the key material needs this handshake. )

Reply:
```
> What is "server enforced PIN protection"?

The mnemonic when generated is immediately encrypted using AES256 using a secret key that is computed through an encrypted and authenticated channel with a [remote] blind oracle server, the authentication involves ephemeral elliptic curve Diffie Hellman exchange with a known server key. The encrypted mnemonic is then stored on the encrypted flash of the Jade and protected by secure boot [v2].

The blind server is not aware of the actual PIN nor can brute force it. The blind server also is not associated with the users account (it is not associated with the per user derivation paths)

If the PIN is entered incorrectly 3 times the server and Jade both delete the secret requiring a from scratch mnemonic restore on Jade.

The companion app nor any ISP/WiFi AP in the middle can inspect the data passed between Jade and the oracle, including being able to tell if the PIN is correct or wrong (other than potential timing attacks). The oracle is available via Tor and open source and in the future users will be able to run their own and point their Jade to it if they prefer.
```

from https://www.reddit.com/r/Bitcoin/comments/kqgehd/were_the_blockstream_team_and_we_just_announced/gi3tvq0?utm_source=share&utm_medium=web2x&context=3